pkg/splunk/enterprise/workerpool.go

```diff
@@ -0,0 +1,191 @@
+	S3Reponse          splclient.S3Response
+	wg                 *sync.WaitGroup
+	availableDiskSpace uint64
+	mu                 sync.Mutex
+}
+
+// NewAppInstallWorkerPool returns a new worker pool
+func NewAppInstallWorkerPool(numOfWorkers uint64, jobs []enterpriseApi.AppDeploymentInfo,
+	s3ClientMgr S3ClientManager, s3Response splclient.S3Response,
+	availableDiskSpace uint64, downloadPath, scope string) *AppInstallWorkerPool {
+	w := &AppInstallWorkerPool{
+		NumOfWorkers:       numOfWorkers,
+		JobList:            getObjectsAsPointers(jobs).([]*enterpriseApi.AppDeploymentInfo),
+		JobsChan:           make(chan *enterpriseApi.AppDeploymentInfo),
+		DownloadPath:       downloadPath,
+		Scope:              scope,
+		S3ClientMgr:        s3ClientMgr,
+		S3Reponse:          s3Response,
+		availableDiskSpace: availableDiskSpace,
+		wg:                 new(sync.WaitGroup),
+	}
+	return w
+}
+
+// Run is the main work that workers do to install the apps.
+// Workers wait for the jobs on the jobs channel and pick them up as they come in
+func (w *AppInstallWorkerPool) Run(id int, jobs <-chan *enterpriseApi.AppDeploymentInfo) {
+	scopedLog := log.WithName("AppInstallWorkerPool.run()").WithValues("worker id", id)
+	for job := range jobs {
+		scopedLog.Info("Started install of app", "appName", job.AppName)
+
+		object := getRemoteObjectFromS3Response(job.AppName, w.S3Reponse)
+		localFile := getLocalAppFileName(w.DownloadPath, job.AppName, *object.Etag)
+		remoteFile := *object.Key
+
+		// download the app from remote storage
+		err := w.S3ClientMgr.DownloadApp(remoteFile, localFile, *object.Etag)
+		if err != nil {
+			scopedLog.Error(err, "unable to download app", "appName", job.AppName)
+			setAppDownloadState(job, enterpriseApi.DownloadError)
+		} else {
+			// download is successfull, update the state
+			setAppDownloadState(job, enterpriseApi.DownloadComplete)
+
+			//TODO: gaurav, need to fill in below code
+			// copy the app to the splunk pod
+
+			// untar it, if it is cluster scoped
+
+			// if everything is successful, delete the app from operator pod
+		}
+
+		scopedLog.Info("finished job")
+		w.mu.Lock()
+		w.availableDiskSpace = w.availableDiskSpace + job.Size
+		scopedLog.Info("availableDiskSpace after freeing up space", "availableDiskSpace", w.availableDiskSpace)
+		w.mu.Unlock()
+
+		w.wg.Done()
+	}
+}
+
+// isAppAlreadyDownloaded checks if the app is already downloaded on the operator pod by checking a couple of things -
+// 1. is the app bundle present on the operator pod &&
+// 2. if the size on disk matches the size reported from remote storage
+func (w *AppInstallWorkerPool) isAppAlreadyDownloaded(app *enterpriseApi.AppDeploymentInfo) bool {
+	scopedLog := log.WithName("isAppAlreadyDownloaded").WithValues("app name", app.AppName)
+	object := getRemoteObjectFromS3Response(app.AppName, w.S3Reponse)
+	localAppFileName := getLocalAppFileName(w.DownloadPath, app.AppName, *object.Etag)
+
+	var fileInfo os.FileInfo
```

Comment:
Just for reference, I got the build issue on CircleCI resolved by using `os.FileInfo` instead of `fs.FileInfo` and hence importing package `os` rather than `io/fs`. So, for older versions of golang(<1.16.x) there was `FileInfo` struct which was implemented in `os` package but in newer versions(>=1.16.x) it is implemented under `io/fs` package. Hence instead of using  `fs.FileInfo` I had to use `os.FileInfo` which is backward compatible too in the sense that for newer versions, FileInfo is mapped to `fs.FileInfo`.